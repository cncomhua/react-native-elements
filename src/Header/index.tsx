import { withTheme } from '../config';
import { Header, HeaderProps } from './Header';

export { Header };
export type { HeaderProps };
/**
 * RNE Header
 *
 * **Preview**
 *
 * ![URL](URL)
 *
 * [**Documentation**](https://reactnativeelements.com/docs/header)
 */
export default withTheme<HeaderProps, {}>(Header, 'Header');
