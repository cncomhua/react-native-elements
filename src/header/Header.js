import React from 'react';
import PropTypes from 'prop-types';
import { StatusBar, View } from 'react-native';
import isEmpty from 'lodash.isempty';
import DummyNavButton from './DummyNavButton';
import NavButton from './NavButton';
import Title from './Title';
import { ViewPropTypes } from '../config';

function generateChild(value, type) {
  if (React.isValidElement(value)) {
    return (
      <View key={type}>
        {value}
      </View>
    );
  } else if (typeof value === 'object' && !isEmpty(value)) {
    return type === 'center'
      ? <Title {...value} key={type} />
      : <NavButton {...value} key={type} />;
  }
  return type === 'center' ? null : <DummyNavButton key={type} />;
}

function populateChildren(propChildren) {
  const childrenArray = [];

  const leftComponent = generateChild(propChildren.leftComponent, 'left');
  const centerComponent = generateChild(propChildren.centerComponent, 'center');
  const rightComponent = generateChild(propChildren.rightComponent, 'right');

  childrenArray.push(leftComponent, centerComponent, rightComponent);

  return childrenArray;
}

class Header extends React.PureComponent {
  render() {
    const {
      children,
      statusBarProps,
      leftComponent,
      centerComponent,
      rightComponent,
      backgroundColor,
      outerContainerStyles,
      innerContainerStyles,
      ...attributes
    } = this.props;

    let propChildren = [];

    if (leftComponent || centerComponent || rightComponent) {
      propChildren = populateChildren({
        leftComponent,
        centerComponent,
        rightComponent,
      });
    }

    const styles = this.context.theme.header;

    return (
      <View
        style={[
          styles.outerContainer,
          { backgroundColor },
          outerContainerStyles,
        ]}
        {...attributes}
      >
        <StatusBar {...statusBarProps} />
        <View style={[styles.innerContainer, innerContainerStyles]}>
          {propChildren.length > 0 ? propChildren : children}
        </View>
      </View>
    );
  }
}

Header.propTypes = {
  leftComponent: PropTypes.object,
  centerComponent: PropTypes.object,
  rightComponent: PropTypes.object,
  backgroundColor: PropTypes.string,
  outerContainerStyles: ViewPropTypes.style,
  innerContainerStyles: ViewPropTypes.style,
  children: PropTypes.element,
  statusBarProps: PropTypes.object,
};

Header.contextTypes = {
  theme: PropTypes.object.isRequired,
};

export default Header;
