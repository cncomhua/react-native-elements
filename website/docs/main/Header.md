---


id: header


title: Header


slug: /header

---



import Usage from './usage/Header/Header.md'



Headers are navigation components that display information and actions relating to the current screen.

**Note:**

Make sure that you have completed [Step 3](getting_started.md#step-3-setup-react-native-safe-area-context) in the setup guide before using `Header`.



## Usage


<Usage />

---


## Props

### Header




## Reference

### Header

