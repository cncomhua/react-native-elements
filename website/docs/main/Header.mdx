---
id: header

title: Header

slug: /header
---

import Usage from './usage/Header/Header.mdx';

import Props from './props/Header.mdx';

import { Header } from 'react-native-elements';

import Tabs from '@theme/Tabs';

import TabItem from '@theme/TabItem';

Headers are navigation components that display information and actions relating to the current screen.

**Note:**

Make sure that you have completed [Step 3](getting_started.md#step-3-setup-react-native-safe-area-context) in the setup guide before using `Header`.

## Usage

<Usage />

## Props

<Props />
