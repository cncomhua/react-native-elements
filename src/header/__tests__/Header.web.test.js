import React from 'react';
import { Text, TouchableHighlight, View } from 'react-native';
import { Button } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const btnCfg = { icon: 'home' };
const titleCfg = { text: 'This is a title' };
import parseSamples, {
  assignRef,
  genRefId,
} from 'enzyme-styleguidist-sample-parser';
import options, {
  snapShot,
  buildJsxForGuideMethod,
  ensureCalled,
  onlyEnsureCalled,
  onlySnapshots,
} from '../../../samples';
import Header from '../Header';
import NavButton from '../NavButton';
import Title from '../Title';

const props = {
  noProps: {
    component: Header,
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  statusBarProps: {
    component: Header,
    props: { statusBarProps: { animated: true } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  leftComponent: {
    component: Header,
    props: { statusBarProps: { animated: true } },
    enzyme: {
      buildJsx: () => {
        return <Header leftComponent={<Text>sample component</Text>} />;
      },
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {
      script: `const Text = RN.Text;\n`,
      buildJsx: () => {
        return <Header leftComponent={<Text>sample component</Text>} />;
      },
    },
  },
  'leftComponent icon': {
    component: Header,
    props: { leftComponent: { icon: 'directions-bike' } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  centerComponent: {
    component: Header,
    props: { statusBarProps: { animated: true } },
    enzyme: {
      buildJsx: () => {
        return <Header centerComponent={<Text>sample component</Text>} />;
      },
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {
      script: `const Text = RN.Text;\n`,
      buildJsx: () => {
        return <Header centerComponent={<Text>sample component</Text>} />;
      },
    },
  },
  'centerComponent text': {
    component: Header,
    props: { centerComponent: { text: 'text sample' } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  rightComponent: {
    component: Header,
    props: { statusBarProps: { animated: true } },
    enzyme: {
      buildJsx: () => {
        return <Header rightComponent={<Text>sample component</Text>} />;
      },
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {
      script: `const Text = RN.Text;\n`,
      buildJsx: () => {
        return <Header rightComponent={<Text>sample component</Text>} />;
      },
    },
  },
  'rightComponent icon': {
    component: Header,
    props: { rightComponent: { icon: 'pets' } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  backgroundColor: {
    component: Header,
    props: { backgroundColor: '#071' },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  outerContainerStyles: {
    component: Header,
    props: { outerContainerStyles: { borderWidth: 5 } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  innerContainerStyles: {
    component: Header,
    props: { innerContainerStyles: { borderWidth: 5 } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
};

const samples = {
  props,
};

parseSamples(
  {
    Misc: { sectionComponents: { Header: { samples } } },
  },
  options
);

describe('Header Component', () => {
  it('should render without issues', () => {
    const component = shallow(<Header />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render children when passed in', () => {
    const component = shallow(
      <Header>
        <Button title="Test button" onPress={() => {}} />
      </Header>
    );

    expect(component.find('Button').length).toBe(1);
  });

  it('should render multiple children when passed in', () => {
    const component = shallow(
      <Header>
        <Button title="Test button 1" onPress={() => {}} />
        <Button title="Test button 2" onPress={() => {}} />
      </Header>
    );

    expect(component.find('Button').length).toBe(2);
  });

  it('should render left component by passing a config through props', () => {
    const component = shallow(<Header leftComponent={btnCfg} />);

    expect(component.find(NavButton).length).toBe(1);
  });

  it('should render left component by passing a component through props', () => {
    const component = shallow(
      <Header
        leftComponent={<Button title="Test button" onPress={() => {}} />}
      />
    );

    expect(component.find('Button').length).toBe(1);
  });

  it('should render right component by passing a config through props', () => {
    const component = shallow(<Header rightComponent={btnCfg} />);

    expect(component.find(NavButton).length).toBe(1);
  });

  it('should render right component by passing a component through props', () => {
    const component = shallow(
      <Header
        rightComponent={<Button title="Test button" onPress={() => {}} />}
      />
    );

    expect(component.find('Button').length).toBe(1);
  });

  it('should render center component by passing a config through props', () => {
    const component = shallow(<Header centerComponent={titleCfg} />);

    expect(component.find(Title).length).toBe(1);
  });

  it('should render center component by passing a component through props', () => {
    const component = shallow(
      <Header
        centerComponent={<Button title="Test button" onPress={() => {}} />}
      />
    );

    expect(component.find('Button').length).toBe(1);
  });

  it('should allow to pass backgroundColor through prop', () => {
    const component = shallow(<Header backgroundColor="#aaa" />);

    expect(
      component
        .find('View')
        .first()
        .props().style[1].backgroundColor
    ).toBe('#aaa');
  });

  it('should allow to pass styles through outerContainerStyles prop', () => {
    const component = shallow(
      <Header outerContainerStyles={{ backgroundColor: '#ccc' }} />
    );

    expect(
      component
        .find('View')
        .at(0)
        .props().style[2].backgroundColor
    ).toBe('#ccc');
  });

  it('should allow to pass styles through innerContainerStyles prop', () => {
    const component = shallow(
      <Header innerContainerStyles={{ backgroundColor: '#ccc' }} />
    );

    expect(
      component
        .find('View')
        .at(1)
        .props().style[1].backgroundColor
    ).toBe('#ccc');
  });

  it('should accept props for StatusBar', () => {
    const component = shallow(<Header statusBarProps={{ hidden: true }} />);

    expect(component.find('StatusBar').props().hidden).toBe(true);
  });
});
