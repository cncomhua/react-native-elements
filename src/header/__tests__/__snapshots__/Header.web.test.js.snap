// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should render without issues 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: backgroundColor ] enzyme-shallow: snapshot: Misc: Header: [ prop: backgroundColor ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#071",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: centerComponent ] enzyme-shallow: snapshot: Misc: Header: [ prop: centerComponent ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <DummyNavButton
      key="left"
    />
    <View
      key="center"
    >
      <Text>
        sample component
      </Text>
    </View>
    <DummyNavButton
      key="right"
    />
  </View>
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: centerComponent text ] enzyme-shallow: snapshot: Misc: Header: [ prop: centerComponent text ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <DummyNavButton
      key="left"
    />
    <Title
      key="center"
      text="text sample"
    />
    <DummyNavButton
      key="right"
    />
  </View>
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: innerContainerStyles ] enzyme-shallow: snapshot: Misc: Header: [ prop: innerContainerStyles ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "borderWidth": 5,
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: leftComponent ] enzyme-shallow: snapshot: Misc: Header: [ prop: leftComponent ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      key="left"
    >
      <Text>
        sample component
      </Text>
    </View>
    <DummyNavButton
      key="right"
    />
  </View>
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: leftComponent icon ] enzyme-shallow: snapshot: Misc: Header: [ prop: leftComponent icon ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <NavButton
      icon="directions-bike"
      key="left"
    />
    <DummyNavButton
      key="right"
    />
  </View>
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: noProps ] enzyme-shallow: snapshot: Misc: Header: [ prop: noProps ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: outerContainerStyles ] enzyme-shallow: snapshot: Misc: Header: [ prop: outerContainerStyles ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "borderWidth": 5,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: rightComponent ] enzyme-shallow: snapshot: Misc: Header: [ prop: rightComponent ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <DummyNavButton
      key="left"
    />
    <View
      key="right"
    >
      <Text>
        sample component
      </Text>
    </View>
  </View>
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: rightComponent icon ] enzyme-shallow: snapshot: Misc: Header: [ prop: rightComponent icon ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <DummyNavButton
      key="left"
    />
    <NavButton
      icon="pets"
      key="right"
    />
  </View>
</View>
`;

exports[`Misc: Header: Props Misc: Header: [ prop: statusBarProps ] enzyme-shallow: snapshot: Misc: Header: [ prop: statusBarProps ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2089dc",
      "borderBottomColor": "#f2f2f2",
      "borderBottomWidth": 1,
      "height": 70,
      "padding": 15,
    }
  }
>
  <StatusBar
    animated={true}
  />
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;
